docs/reference/isr_rules.rst

```diff
@@ -0,0 +1,270 @@
+received to an integer representing total bytes ready for processing. The main program reads the number of bytes,
+processes the bytes, then clears down the number of bytes ready. This will work until an interrupt occurs just
+after the main program has read the number of bytes. The ISR puts the added data into the buffer and updates
+the number received, but the main program has already read the number, so processes the data originally received.
+The newly arrived bytes are lost.
+
+There are various ways of avoiding this hazard, the simplest being to use a circular buffer. If it is not possible
+to use a structure with inherent thread safety other ways are described below.
+
+Reentrancy
+~~~~~~~~~~
+
+A potential hazard may occur if a function or method is shared between the main program and one or more ISR's or
+between multiple ISR's. The issue here is that the function may itself be interrupted and a further instance of
+that function run. If this is to occur, the function must be designed to be reentrant. How this is done is an
+advanced topic beyond the scope of this tutorial.
+
+.. _Critical:
+
+Critical Sections
+~~~~~~~~~~~~~~~~~
+
+An example of a critical section of code is one which accesses more than one variable which can be affected by an ISR. If
+the interrupt happens to occur between accesses to the individual variables, their values will be inconsistent. This is
+an instance of a hazard known as a race condition: the ISR and the main program loop race to alter the variables. To
+avoid inconsistency a means must be employed to ensure that the ISR does not alter the values for the duration of
+the critical section. One way to achieve this is to issue ``pyb.disable_irq()`` before the start of the section, and
+``pyb.enable_irq()`` at the end. Here is an example of this approach:
+
+.. code:: python
+
+    import pyb, micropython, array
+    micropython.alloc_emergency_exception_buf(100)
+
+    class BoundsException(Exception):
+        pass
+
+    ARRAYSIZE = const(20)
+    index = 0
+    data = array.array('i', 0 for x in range(ARRAYSIZE))
+
+    def callback1(t):
+        global data, index
+        for x in range(5):
+            data[index] = pyb.rng() # simulate input
+            index += 1
+            if index >= ARRAYSIZE:
+                raise BoundsException('Array bounds exceeded')
+
+    tim4 = pyb.Timer(4, freq=100, callback = callback1)
+
+    for loop in range(1000):
+        if index > 0:
+            irq_state = pyb.disable_irq()
+            for x in range(index):
+                print(data[x])
+            index = 0
+            pyb.enable_irq(irq_state)
+            print('loop {}'.format(loop))
+        pyb.delay(1)
+
+    tim4.callback(None)
+
+A critical section can comprise a single line of code and a single variable.
+`This <https://github.com/dhylands/upy-examples/blob/master/atomic.py>`_ example illustrates a subtle source
```

Comment:
Let's try not to link to external entities.  I think the only thing relevant here is the `t.counter+=1` piece of code.